File: src/backend/app/transcript.py
from __future__ import annotations

from pathlib import Path
from typing import Iterable

from .models import Conversation, Message, Note

# Single source of truth for transcript header (static part after title). Also served to frontend.
_CONFIG_DIR = Path(__file__).resolve().parent / "config"
_TRANSCRIPT_HEADER_FILE = _CONFIG_DIR / "transcript_header.txt"

_DEFAULT_HEADER_STATIC = """
You are given a structured conversation transcript.

The transcript consists of:
- <<<USER>>> blocks (user messages)
- <<<LLM>>> blocks (assistant responses)
- <<<NOTE>>> blocks (user-authored state notes)

NOTES are contextual clarifications or decisions and must be treated as part of the conversation state.

Continue the conversation by responding as the LLM.
Output only your next single reply (plain text). Do not reproduce wrapper tags.
Do not output further USER, NOTE, or LLM turns—only one assistant reply.

""".strip()


def get_transcript_header_static() -> str:
    """Return the static part of the transcript header (instructions only). Used by BE and served to FE."""
    if _TRANSCRIPT_HEADER_FILE.is_file():
        return _TRANSCRIPT_HEADER_FILE.read_text(encoding="utf-8").strip()
    return _DEFAULT_HEADER_STATIC


WRAPPER_USER_START = "<<<USER>>>"
WRAPPER_USER_END = "<<<END USER>>>"
WRAPPER_LLM_START = "<<<LLM>>>"
WRAPPER_LLM_END = "<<<END LLM>>>"
WRAPPER_NOTE_START = "<<<NOTE>>>"
WRAPPER_NOTE_END = "<<<END NOTE>>>"


def _escape_wrappers(text: str) -> str:
    """
    Escape wrapper tags inside message or note content.

    Simple implementation for Phase 1: replace the opening sequence
    with a visually similar but different token.
    """
    return (
        text.replace("<<<USER>>>", "<< <USER>>")
        .replace("<<<END USER>>>", "<< <END USER>>")
        .replace("<<<LLM>>>", "<< <LLM>>")
        .replace("<<<END LLM>>>", "<< <END LLM>>")
        .replace("<<<NOTE>>>", "<< <NOTE>>")
        .replace("<<<END NOTE>>>", "<< <END NOTE>>")
    )


def build_transcript_header(conversation: Conversation) -> str:
    title = conversation.title or "Conversation"
    static = get_transcript_header_static()
    return f"{title}\n\n{static}"


def serialize_message_block(message: Message, notes: Iterable[Note]) -> str:
    if message.role == "user":
        start, end = WRAPPER_USER_START, WRAPPER_USER_END
    else:
        start, end = WRAPPER_LLM_START, WRAPPER_LLM_END

    parts: list[str] = [
        start,
        _escape_wrappers(message.content),
        end,
        "",
    ]

    sorted_notes = sorted(notes, key=lambda n: (n.created_at, str(n.id)))
    for note in sorted_notes:
        parts.extend(
            [
                WRAPPER_NOTE_START,
                _escape_wrappers(note.content),
                WRAPPER_NOTE_END,
                "",
            ]
        )

    return "\n".join(parts).rstrip()  # avoid trailing spaces


def build_transcript_for_path(
    conversation: Conversation, path_messages: list[Message], notes_by_message: dict
) -> str:
    """
    Build canonical transcript for root→active path.

    - Messages serialized in order
    - Notes inserted immediately after their host message
    - Exactly one blank line between blocks (handled by join logic)
    """
    header = build_transcript_header(conversation)
    blocks: list[str] = [header]

    for msg in path_messages:
        msg_notes = notes_by_message.get(msg.id, [])
        blocks.append(serialize_message_block(msg, msg_notes))

    return "\n\n".join(blocks).strip() + "\n"

File: src/frontend/src/transcript.ts
import type { ConversationTreeResponse, MessageNode, NoteItem, UUID } from "./api/types";
import { pathFromRoot } from "./components/treeLayout";

const WRAPPER_USER_START = "<<<USER>>>";
const WRAPPER_USER_END = "<<<END USER>>>";
const WRAPPER_LLM_START = "<<<LLM>>>";
const WRAPPER_LLM_END = "<<<END LLM>>>";
const WRAPPER_NOTE_START = "<<<NOTE>>>";
const WRAPPER_NOTE_END = "<<<END NOTE>>>";

function escapeWrappers(text: string): string {
  return text
    .replace(/<<<USER>>>/g, "<< <USER>>")
    .replace(/<<<END USER>>>/g, "<< <END USER>>")
    .replace(/<<<LLM>>>/g, "<< <LLM>>")
    .replace(/<<<END LLM>>>/g, "<< <END LLM>>")
    .replace(/<<<NOTE>>>/g, "<< <NOTE>>")
    .replace(/<<<END NOTE>>>/g, "<< <END NOTE>>");
}

function buildHeader(conversationTitle: string | null): string {
  const title = conversationTitle?.trim() || "Conversation";
  return [
    title,
    "",
    "You are given a structured conversation transcript.",
    "",
    "The transcript consists of:",
    "- <<<USER>>> blocks (user messages)",
    "- <<<LLM>>> blocks (assistant responses)",
    "- <<<NOTE>>> blocks (user-authored state notes)",
    "",
    "NOTES are contextual clarifications or decisions and must be treated as part of the conversation state.",
    "",
    "Continue the conversation by responding as the LLM.",
    "Do not reproduce wrapper tags in your response.",
    "",
  ].join("\n");
}

function serializeMessageBlock(message: MessageNode, notes: NoteItem[]): string {
  const start = message.role === "user" ? WRAPPER_USER_START : WRAPPER_LLM_START;
  const end = message.role === "user" ? WRAPPER_USER_END : WRAPPER_LLM_END;
  const parts: string[] = [start, escapeWrappers(message.content), end, ""];
  const sortedNotes = [...notes].sort(
    (a, b) => a.created_at.localeCompare(b.created_at) || a.id.localeCompare(b.id)
  );
  for (const note of sortedNotes) {
    parts.push(WRAPPER_NOTE_START, escapeWrappers(note.content), WRAPPER_NOTE_END, "");
  }
  return parts.join("\n").replace(/\n+$/, "");
}

/**
 * Build full transcript for root→active path from tree (messages + notes).
 * Used when loading a conversation or when needs_context_rebuild (refetch tree then build).
 */
export function buildTranscriptFromTree(tree: ConversationTreeResponse): string {
  const path = pathFromRoot(tree.messages, tree.active_state.active_message_id);
  const title = tree.conversation_title ?? null;
  const header = buildHeader(title);
  const blocks: string[] = [header];
  for (const msg of path) {
    const notes = tree.notes[msg.id] ?? [];
    blocks.push(serializeMessageBlock(msg, notes));
  }
  return blocks.join("\n\n").trim() + "\n";
}

/**
 * Build initial transcript for create conversation: header + first user block.
 */
export function buildInitialTranscript(
  conversationTitle: string | null,
  firstMessageContent: string
): string {
  const header = buildHeader(conversationTitle);
  const userBlock = [
    WRAPPER_USER_START,
    escapeWrappers(firstMessageContent),
    WRAPPER_USER_END,
  ].join("\n");
  return header + "\n\n" + userBlock + "\n";
}

/**
 * Append a new user message block to the transcript (for send).
 */
export function appendUserBlock(transcript: string, content: string, _author?: string): string {
  const block = [WRAPPER_USER_START, escapeWrappers(content), WRAPPER_USER_END].join("\n");
  return transcript.trimEnd() + "\n\n" + block + "\n";
}
